### DISCUSSION TITLE
Write unit test cases for activities in ELSA v3

### DISCUSSION THREAD

**Discussion by manaliadhya-302** [elsa-workflows/elsa-core#4249] [Q&A]
How to write unit test cases for activities in ELSA v3? 
I am using ELSA v3. I have gone through the v3 branch of 'elsa-core' repo, but didn't see any unit test for stand-alone activities. 
What is the plan for it?
@sfmskywalker

**Comment by sfmskywalker** [comment]
I’ll be adding activity units tests at some point.

**Reply by abhjitsantra** [reply]
@sfmskywalker, do we have sample activity unit tests to refer yet?

**Reply by sfmskywalker** [reply]
Yes, for now we have two very basic tests for two basic activities, as a starting point: https://github.com/elsa-workflows/elsa-core/tree/v3/test/unit/Elsa.Activities.UnitTests

Activities require an execution context to run in. To simplify setting up this context, the unit tests construct an `IWorkflowRunner`, which is capable of executing any `IActivity` and will construct the necessary activity execution context.

This is is simplified even further with the `RunActivityAsync` extension method on the service provider, allowing you to do this to run an activity:

```csharp
// Instantiate the activity.
var writeLine = new WriteLine("Hello world!");

// Run the activity.
await _serviceProvider.RunActivityAsync(writeLine);
```

Here's a complete unit test:

```csharp
using Elsa.Testing.Shared;
using Elsa.Workflows.Core.Activities;
using Xunit.Abstractions;

namespace Elsa.Activities.UnitTests;

public class WriteLineTests
{
    private readonly CapturingTextWriter _capturingTextWriter = new();
    private readonly IServiceProvider _serviceProvider;

    public WriteLineTests(ITestOutputHelper testOutputHelper)
    {
        _serviceProvider = new TestApplicationBuilder(testOutputHelper).WithCapturingTextWriter(_capturingTextWriter).Build();
    }
    
    [Fact(DisplayName = "WriteLine prints the expected line to the console.")]
    public async Task Test1()
    {
        const string expectedLine = "Hello world!";
        var writeLine = new WriteLine(expectedLine);
        await _serviceProvider.RunActivityAsync(writeLine);
        Assert.Equal(expectedLine, _capturingTextWriter.Lines.Single());
    }
}
```

**Reply by abhjitsantra** [reply]
It seems the example you shared is to write tests for very basic activities having only primitive datatype inputs. 
I am trying to write unit tests for an activity that has `Variable` type input property. I tried creating a variable, registering it and then executing the activity like the example below but the activity can't get the variable from the `ActivityExecutionContext` and always results to `null`.

```CSharp
Variable date = new ("date", "12-12-2023");
var memoryRegister = new MemoryRegister(new Dictionary<string, MemoryBlock>
{
    { date.Id, new MemoryBlock(date.Value, new VariableBlockMetadata(date, typeof(object), true)) }
});

var context = new ExpressionExecutionContext(_serviceProvider, memoryRegister);

CustomActivity activity = new CustomActivity()
{
    Date = date,
    Days = new ("5")
};

var result = await context.ServiceProvider.RunActivityAsync(activity);
```

Do I need to register the variable in a different way to make this working?
